# Springboard Data Science Career Track Unit 4 Challenge - Tier 3 Complete

Objective
Which boroughs of London have seen the greatest increase in housing prices, on average, over the last two decades?

### 1. Sourcing and Loading 

Any Data Science project kicks off by importing  ***pandas***. The documentation of this wonderful library can be found [here](https://pandas.pydata.org/). As you've seen, pandas is conveniently connected to the [Numpy](http://www.numpy.org/) and [Matplotlib](https://matplotlib.org/) libraries. 

***Hint:*** This part of the data science pipeline will test those skills you acquired in the pandas Foundations course, Module One. 

#### 1.1. Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#### 1.2.  Loading the data
Your data comes from the [London Datastore](https://data.london.gov.uk/): a free, open-source data-sharing portal for London-oriented datasets. 

In [2]:
#create variable from data source
url_LondonHousePrices = "https://data.london.gov.uk/download/uk-house-price-index/70ac0766-8902-4eb5-aab5-01951aaed773/UK%20House%20price%20index.xls"

# create dataframe for Excel sheet that contains the Average prices of the houses 
properties = pd.read_excel(url_LondonHousePrices, sheet_name='Average price', index_col= None)

### 2. Cleaning, transforming, and visualizing
This second stage is arguably the most important part of any Data Science project. The first thing to do is take a proper look at the data. Cleaning forms the majority of this stage, and can be done both before or after Transformation.

The end goal of data cleaning is to have tidy data. When data is tidy: 

1. Each variable has a column.
2. Each observation forms a row.

Keep the end goal in mind as you move through this process, every step will take you closer. 



***Hint:*** This part of the data science pipeline should test those skills you acquired in: 
- Intermediate Python for data science, all modules.
- pandas Foundations, all modules. 
- Manipulating DataFrames with pandas, all modules.
- Data Types for Data Science, Module Four.
- Python Data Science Toolbox - Part One, all modules

**2.1. Exploring your data** 

Think about your pandas functions for checking out a dataframe. 

In [3]:
properties.shape

(322, 49)

In [4]:
properties.head()

Unnamed: 0 City of London Barking & Dagenham       Barnet       Bexley  \
0        NaT      E09000001          E09000002    E09000003    E09000004   
1 1995-01-01    91448.98487         50460.2266  93284.51832  64958.09036   
2 1995-02-01    82202.77314        51085.77983  93190.16963  64787.92069   
3 1995-03-01    79120.70256        51268.96956  92247.52435  64367.49344   
4 1995-04-01    77101.20804        53133.50526  90762.87492  64277.66881   

         Brent      Bromley       Camden      Croydon       Ealing  ...  \
0    E09000005    E09000006    E09000007    E09000008    E09000009  ...   
1  71306.56698  81671.47692  120932.8881  69158.16225  79885.89069  ...   
2  72022.26197  81657.55944  119508.8622  68951.09542  80897.06551  ...   
3  72015.76274  81449.31143  120282.2131  68712.44341  81379.86288  ...   
4  72965.63094  81124.41227   120097.899  68610.04641  82188.90498  ...   

    NORTH WEST YORKS & THE HUMBER EAST MIDLANDS WEST MIDLANDS EAST OF ENGLAND  \
0    E12000002          E12000003     E12000004     E12000005       E12000006   
1  43958.48001        44803.42878   45544.52227   48527.52339      56701.5961   
2  43925.42289        44528.80721   46051.57066   49341.29029     56593.59475   
3   44434.8681        45200.46775   45383.82395   49442.17973     56171.18278   
4   44267.7796        45614.34341   46124.23045   49455.93299     56567.89582   

        LONDON   SOUTH EAST   SOUTH WEST Unnamed: 47      England  
0    E12000007    E12000008    E12000009         NaN    E92000001  
1  74435.76052  64018.87894   54705.1579         NaN  53202.77128  
2  72777.93709  63715.02399  54356.14843         NaN   53096.1549  
3  73896.84204  64113.60858  53583.07667         NaN   53201.2843  
4  74455.28754  64623.22395  54786.01938         NaN   53590.8548  

[5 rows x 49 columns]

**2.2. Cleaning the data**

You might find you need to transpose your dataframe, check out what its row indexes are, and reset the index. You  also might find you need to assign the values of the first row to your column headings  . (Hint: recall the .columns feature of DataFrames, as well as the iloc[] method).

Don't be afraid to use StackOverflow for help  with this.

In [5]:
#transpose dataframe to switch index and columns
properties_T=properties.transpose()

In [6]:
properties_T.head()

0                    1                    2    \
Unnamed: 0                NaT  1995-01-01 00:00:00  1995-02-01 00:00:00   
City of London      E09000001          91448.98487          82202.77314   
Barking & Dagenham  E09000002           50460.2266          51085.77983   
Barnet              E09000003          93284.51832          93190.16963   
Bexley              E09000004          64958.09036          64787.92069   

                                    3                    4    \
Unnamed: 0          1995-03-01 00:00:00  1995-04-01 00:00:00   
City of London              79120.70256          77101.20804   
Barking & Dagenham          51268.96956          53133.50526   
Barnet                      92247.52435          90762.87492   
Bexley                      64367.49344          64277.66881   

                                    5                    6    \
Unnamed: 0          1995-05-01 00:00:00  1995-06-01 00:00:00   
City of London              84409.14932          94900.51244   
Barking & Dagenham          53042.24852          53700.34831   
Barnet                      90258.00033          90107.23471   
Bexley                      63997.13588          64252.32335   

                                    7                    8    \
Unnamed: 0          1995-07-01 00:00:00  1995-08-01 00:00:00   
City of London              110128.0423          112329.4376   
Barking & Dagenham          52113.12157          52232.19868   
Barnet                      91441.24768          92361.31512   
Bexley                      63722.70055          64432.60005   

                                    9    ...                  312  \
Unnamed: 0          1995-09-01 00:00:00  ...  2020-12-01 00:00:00   
City of London              104473.1096  ...          801999.0466   
Barking & Dagenham          51471.61353  ...          309560.1023   
Barnet                      93273.12245  ...          532528.0926   
Bexley                      64509.54767  ...          352260.8385   

                                    313                  314  \
Unnamed: 0          2021-01-01 00:00:00  2021-02-01 00:00:00   
City of London              754194.0982          725011.1264   
Barking & Dagenham          312470.4977          315749.3442   
Barnet                      538683.8133          537171.1625   
Bexley                      355756.7322          359323.8616   

                                    315                  316  \
Unnamed: 0          2021-03-01 00:00:00  2021-04-01 00:00:00   
City of London              719703.4989          728496.7198   
Barking & Dagenham          312381.5975          313131.2516   
Barnet                      537972.6322           530778.403   
Bexley                      361782.2198          362053.2197   

                                    317                  318  \
Unnamed: 0          2021-05-01 00:00:00  2021-06-01 00:00:00   
City of London              720954.3631          729878.6878   
Barking & Dagenham          312430.6526          317234.7354   
Barnet                      533319.3495          541296.7827   
Bexley                      361919.1777          364464.1865   

                                    319                  320  \
Unnamed: 0          2021-07-01 00:00:00  2021-08-01 00:00:00   
City of London                        -                    -   
Barking & Dagenham          319549.9296          322496.3752   
Barnet                      536012.6687          544558.8072   
Bexley                      366451.6388          373547.2951   

                                    321  
Unnamed: 0          2021-09-01 00:00:00  
City of London                        -  
Barking & Dagenham          329287.2922  
Barnet                       533093.293  
Bexley                       375156.155  

[5 rows x 322 columns]

In [7]:
#check the indexs for unique IDs
properties_T.index

Index(['Unnamed: 0', 'City of London', 'Barking & Dagenham', 'Barnet',
       'Bexley', 'Brent', 'Bromley', 'Camden', 'Croydon', 'Ealing', 'Enfield',
       'Greenwich', 'Hackney', 'Hammersmith & Fulham', 'Haringey', 'Harrow',
       'Havering', 'Hillingdon', 'Hounslow', 'Islington',
       'Kensington & Chelsea', 'Kingston upon Thames', 'Lambeth', 'Lewisham',
       'Merton', 'Newham', 'Redbridge', 'Richmond upon Thames', 'Southwark',
       'Sutton', 'Tower Hamlets', 'Waltham Forest', 'Wandsworth',
       'Westminster', 'Unnamed: 34', 'Inner London', 'Outer London',
       'Unnamed: 37', 'NORTH EAST', 'NORTH WEST', 'YORKS & THE HUMBER',
       'EAST MIDLANDS', 'WEST MIDLANDS', 'EAST OF ENGLAND', 'LONDON',
       'SOUTH EAST', 'SOUTH WEST', 'Unnamed: 47', 'England'],
      dtype='object')

In [8]:
#reset index to assign unique IDs
properties_T = properties_T.reset_index()

In [9]:
#check index
properties_T.index 

RangeIndex(start=0, stop=49, step=1)

In [10]:
#view index
properties_T.head()

index          0                    1                    2  \
0          Unnamed: 0        NaT  1995-01-01 00:00:00  1995-02-01 00:00:00   
1      City of London  E09000001          91448.98487          82202.77314   
2  Barking & Dagenham  E09000002           50460.2266          51085.77983   
3              Barnet  E09000003          93284.51832          93190.16963   
4              Bexley  E09000004          64958.09036          64787.92069   

                     3                    4                    5  \
0  1995-03-01 00:00:00  1995-04-01 00:00:00  1995-05-01 00:00:00   
1          79120.70256          77101.20804          84409.14932   
2          51268.96956          53133.50526          53042.24852   
3          92247.52435          90762.87492          90258.00033   
4          64367.49344          64277.66881          63997.13588   

                     6                    7                    8  ...  \
0  1995-06-01 00:00:00  1995-07-01 00:00:00  1995-08-01 00:00:00  ...   
1          94900.51244          110128.0423          112329.4376  ...   
2          53700.34831          52113.12157          52232.19868  ...   
3          90107.23471          91441.24768          92361.31512  ...   
4          64252.32335          63722.70055          64432.60005  ...   

                   312                  313                  314  \
0  2020-12-01 00:00:00  2021-01-01 00:00:00  2021-02-01 00:00:00   
1          801999.0466          754194.0982          725011.1264   
2          309560.1023          312470.4977          315749.3442   
3          532528.0926          538683.8133          537171.1625   
4          352260.8385          355756.7322          359323.8616   

                   315                  316                  317  \
0  2021-03-01 00:00:00  2021-04-01 00:00:00  2021-05-01 00:00:00   
1          719703.4989          728496.7198          720954.3631   
2          312381.5975          313131.2516          312430.6526   
3          537972.6322           530778.403          533319.3495   
4          361782.2198          362053.2197          361919.1777   

                   318                  319                  320  \
0  2021-06-01 00:00:00  2021-07-01 00:00:00  2021-08-01 00:00:00   
1          729878.6878                    -                    -   
2          317234.7354          319549.9296          322496.3752   
3          541296.7827          536012.6687          544558.8072   
4          364464.1865          366451.6388          373547.2951   

                   321  
0  2021-09-01 00:00:00  
1                    -  
2          329287.2922  
3           533093.293  
4           375156.155  

[5 rows x 323 columns]

In [11]:
#confirm columns are integers
properties_T.columns

Index(['index',       0,       1,       2,       3,       4,       5,       6,
             7,       8,
       ...
           312,     313,     314,     315,     316,     317,     318,     319,
           320,     321],
      dtype='object', length=323)

In [12]:
#see the row at index 0. 
properties_T.iloc[[0]]

index    0                    1                    2  \
0  Unnamed: 0  NaT  1995-01-01 00:00:00  1995-02-01 00:00:00   

                     3                    4                    5  \
0  1995-03-01 00:00:00  1995-04-01 00:00:00  1995-05-01 00:00:00   

                     6                    7                    8  ...  \
0  1995-06-01 00:00:00  1995-07-01 00:00:00  1995-08-01 00:00:00  ...   

                   312                  313                  314  \
0  2020-12-01 00:00:00  2021-01-01 00:00:00  2021-02-01 00:00:00   

                   315                  316                  317  \
0  2021-03-01 00:00:00  2021-04-01 00:00:00  2021-05-01 00:00:00   

                   318                  319                  320  \
0  2021-06-01 00:00:00  2021-07-01 00:00:00  2021-08-01 00:00:00   

                   321  
0  2021-09-01 00:00:00  

[1 rows x 323 columns]

In [13]:
# reassign the columns of dataFrame the values within this row at index 0.
properties_T.columns = properties_T.iloc[0]

In [14]:
#check dataframe again
properties_T.head()

0          Unnamed: 0        NaN  1995-01-01 00:00:00  1995-02-01 00:00:00  \
0          Unnamed: 0        NaT  1995-01-01 00:00:00  1995-02-01 00:00:00   
1      City of London  E09000001          91448.98487          82202.77314   
2  Barking & Dagenham  E09000002           50460.2266          51085.77983   
3              Barnet  E09000003          93284.51832          93190.16963   
4              Bexley  E09000004          64958.09036          64787.92069   

0  1995-03-01 00:00:00  1995-04-01 00:00:00  1995-05-01 00:00:00  \
0  1995-03-01 00:00:00  1995-04-01 00:00:00  1995-05-01 00:00:00   
1          79120.70256          77101.20804          84409.14932   
2          51268.96956          53133.50526          53042.24852   
3          92247.52435          90762.87492          90258.00033   
4          64367.49344          64277.66881          63997.13588   

0  1995-06-01 00:00:00  1995-07-01 00:00:00  1995-08-01 00:00:00  ...  \
0  1995-06-01 00:00:00  1995-07-01 00:00:00  1995-08-01 00:00:00  ...   
1          94900.51244          110128.0423          112329.4376  ...   
2          53700.34831          52113.12157          52232.19868  ...   
3          90107.23471          91441.24768          92361.31512  ...   
4          64252.32335          63722.70055          64432.60005  ...   

0  2020-12-01 00:00:00  2021-01-01 00:00:00  2021-02-01 00:00:00  \
0  2020-12-01 00:00:00  2021-01-01 00:00:00  2021-02-01 00:00:00   
1          801999.0466          754194.0982          725011.1264   
2          309560.1023          312470.4977          315749.3442   
3          532528.0926          538683.8133          537171.1625   
4          352260.8385          355756.7322          359323.8616   

0  2021-03-01 00:00:00  2021-04-01 00:00:00  2021-05-01 00:00:00  \
0  2021-03-01 00:00:00  2021-04-01 00:00:00  2021-05-01 00:00:00   
1          719703.4989          728496.7198          720954.3631   
2          312381.5975          313131.2516          312430.6526   
3          537972.6322           530778.403          533319.3495   
4          361782.2198          362053.2197          361919.1777   

0  2021-06-01 00:00:00  2021-07-01 00:00:00  2021-08-01 00:00:00  \
0  2021-06-01 00:00:00  2021-07-01 00:00:00  2021-08-01 00:00:00   
1          729878.6878                    -                    -   
2          317234.7354          319549.9296          322496.3752   
3          541296.7827          536012.6687          544558.8072   
4          364464.1865          366451.6388          373547.2951   

0  2021-09-01 00:00:00  
0  2021-09-01 00:00:00  
1                    -  
2          329287.2922  
3           533093.293  
4           375156.155  

[5 rows x 323 columns]

In [15]:
#drop row at index 0 as it contains no data
properties_T = properties_T.drop(0)

In [16]:
#check datafram header again
properties_T.head()

0          Unnamed: 0        NaN 1995-01-01 00:00:00 1995-02-01 00:00:00  \
1      City of London  E09000001         91448.98487         82202.77314   
2  Barking & Dagenham  E09000002          50460.2266         51085.77983   
3              Barnet  E09000003         93284.51832         93190.16963   
4              Bexley  E09000004         64958.09036         64787.92069   
5               Brent  E09000005         71306.56698         72022.26197   

0 1995-03-01 00:00:00 1995-04-01 00:00:00 1995-05-01 00:00:00  \
1         79120.70256         77101.20804         84409.14932   
2         51268.96956         53133.50526         53042.24852   
3         92247.52435         90762.87492         90258.00033   
4         64367.49344         64277.66881         63997.13588   
5         72015.76274         72965.63094         73704.04743   

0 1995-06-01 00:00:00 1995-07-01 00:00:00 1995-08-01 00:00:00  ...  \
1         94900.51244         110128.0423         112329.4376  ...   
2         53700.34831         52113.12157         52232.19868  ...   
3         90107.23471         91441.24768         92361.31512  ...   
4         64252.32335         63722.70055         64432.60005  ...   
5         74310.48167         74127.03788          73547.0411  ...   

0 2020-12-01 00:00:00 2021-01-01 00:00:00 2021-02-01 00:00:00  \
1         801999.0466         754194.0982         725011.1264   
2         309560.1023         312470.4977         315749.3442   
3         532528.0926         538683.8133         537171.1625   
4         352260.8385         355756.7322         359323.8616   
5         495066.6459         514525.2307         516500.5723   

0 2021-03-01 00:00:00 2021-04-01 00:00:00 2021-05-01 00:00:00  \
1         719703.4989         728496.7198         720954.3631   
2         312381.5975         313131.2516         312430.6526   
3         537972.6322          530778.403         533319.3495   
4         361782.2198         362053.2197         361919.1777   
5         520406.5733         502074.0621         506917.3065   

0 2021-06-01 00:00:00 2021-07-01 00:00:00 2021-08-01 00:00:00  \
1         729878.6878                   -                   -   
2         317234.7354         319549.9296         322496.3752   
3         541296.7827         536012.6687         544558.8072   
4         364464.1865         366451.6388         373547.2951   
5         514055.8987         528684.0547         533696.3904   

0 2021-09-01 00:00:00  
1                   -  
2         329287.2922  
3          533093.293  
4          375156.155  
5         542095.3343  

[5 rows x 323 columns]

**2.3. Cleaning the data (part 2)**

You might we have to **rename** a couple columns. How do you do this? The clue's pretty bold...

In [17]:
#rename Unnamed columns and Nan columns 
properties_T = properties_T.rename(columns = {'Unnamed: 0':'London_Borough', pd.NaT: 'ID'})

In [18]:
#check rename
properties_T.head()

0      London_Borough         ID 1995-01-01 00:00:00 1995-02-01 00:00:00  \
1      City of London  E09000001         91448.98487         82202.77314   
2  Barking & Dagenham  E09000002          50460.2266         51085.77983   
3              Barnet  E09000003         93284.51832         93190.16963   
4              Bexley  E09000004         64958.09036         64787.92069   
5               Brent  E09000005         71306.56698         72022.26197   

0 1995-03-01 00:00:00 1995-04-01 00:00:00 1995-05-01 00:00:00  \
1         79120.70256         77101.20804         84409.14932   
2         51268.96956         53133.50526         53042.24852   
3         92247.52435         90762.87492         90258.00033   
4         64367.49344         64277.66881         63997.13588   
5         72015.76274         72965.63094         73704.04743   

0 1995-06-01 00:00:00 1995-07-01 00:00:00 1995-08-01 00:00:00  ...  \
1         94900.51244         110128.0423         112329.4376  ...   
2         53700.34831         52113.12157         52232.19868  ...   
3         90107.23471         91441.24768         92361.31512  ...   
4         64252.32335         63722.70055         64432.60005  ...   
5         74310.48167         74127.03788          73547.0411  ...   

0 2020-12-01 00:00:00 2021-01-01 00:00:00 2021-02-01 00:00:00  \
1         801999.0466         754194.0982         725011.1264   
2         309560.1023         312470.4977         315749.3442   
3         532528.0926         538683.8133         537171.1625   
4         352260.8385         355756.7322         359323.8616   
5         495066.6459         514525.2307         516500.5723   

0 2021-03-01 00:00:00 2021-04-01 00:00:00 2021-05-01 00:00:00  \
1         719703.4989         728496.7198         720954.3631   
2         312381.5975         313131.2516         312430.6526   
3         537972.6322          530778.403         533319.3495   
4         361782.2198         362053.2197         361919.1777   
5         520406.5733         502074.0621         506917.3065   

0 2021-06-01 00:00:00 2021-07-01 00:00:00 2021-08-01 00:00:00  \
1         729878.6878                   -                   -   
2         317234.7354         319549.9296         322496.3752   
3         541296.7827         536012.6687         544558.8072   
4         364464.1865         366451.6388         373547.2951   
5         514055.8987         528684.0547         533696.3904   

0 2021-09-01 00:00:00  
1                   -  
2         329287.2922  
3          533093.293  
4          375156.155  
5         542095.3343  

[5 rows x 323 columns]

In [19]:
#check how many columns
properties_T.columns

Index([   'London_Borough',                'ID', 1995-01-01 00:00:00,
       1995-02-01 00:00:00, 1995-03-01 00:00:00, 1995-04-01 00:00:00,
       1995-05-01 00:00:00, 1995-06-01 00:00:00, 1995-07-01 00:00:00,
       1995-08-01 00:00:00,
       ...
       2020-12-01 00:00:00, 2021-01-01 00:00:00, 2021-02-01 00:00:00,
       2021-03-01 00:00:00, 2021-04-01 00:00:00, 2021-05-01 00:00:00,
       2021-06-01 00:00:00, 2021-07-01 00:00:00, 2021-08-01 00:00:00,
       2021-09-01 00:00:00],
      dtype='object', name=0, length=323)

**2.4.Transforming the data**

Remember what Wes McKinney said about tidy data? 

You might need to **melt** your DataFrame here. 

In [20]:
# melt column values along the column headings into a single column 
#representing the average house price within a given month and a given borough
clean_properties = pd.melt(properties_T, id_vars= ['London_Borough', 'ID'])

In [21]:
#check melted dataframe
clean_properties.head()

London_Borough         ID          0        value
0      City of London  E09000001 1995-01-01  91448.98487
1  Barking & Dagenham  E09000002 1995-01-01   50460.2266
2              Barnet  E09000003 1995-01-01  93284.51832
3              Bexley  E09000004 1995-01-01  64958.09036
4               Brent  E09000005 1995-01-01  71306.56698

Remember to make sure your column data types are all correct. Average prices, for example, should be floating point numbers... 

In [22]:
#rename columns for better understanding of datapoints
clean_properties = clean_properties.rename(columns = {0: 'Month', 'value': 'Average_price'})

In [23]:
#check renamed columns
clean_properties.head()

London_Borough         ID      Month Average_price
0      City of London  E09000001 1995-01-01   91448.98487
1  Barking & Dagenham  E09000002 1995-01-01    50460.2266
2              Barnet  E09000003 1995-01-01   93284.51832
3              Bexley  E09000004 1995-01-01   64958.09036
4               Brent  E09000005 1995-01-01   71306.56698

In [24]:
#check data types and make sure Average prices are floating point numbers
clean_properties.dtypes

London_Borough            object
ID                        object
Month             datetime64[ns]
Average_price             object
dtype: object

In [25]:
#Call the to_numeric() method on pd, pass the 'Average_price' column into its brackets 
#reassign the result to the clean_properties 'Average_price' column.
clean_properties['Average_price'] = pd.to_numeric(clean_properties['Average_price'], errors='coerce')

In [26]:
#check that Average price data type is corrected
clean_properties.dtypes

London_Borough            object
ID                        object
Month             datetime64[ns]
Average_price            float64
dtype: object

In [27]:
#check for missing values
clean_properties.count()

London_Borough    15408
ID                14445
Month             15408
Average_price     14442
dtype: int64

**2.5. Cleaning the data (part 3)**

Do we have an equal number of observations in the ID, Average Price, Month, and London Borough columns? Remember that there are only 32 London Boroughs. How many entries do you have in that column? 

Check out the contents of the London Borough column, and if you find null values, get rid of them however you see fit. 

In [28]:
#check unique London Borough columns to find any that don't belong  
clean_properties['London_Borough'].unique()

array(['City of London', 'Barking & Dagenham', 'Barnet', 'Bexley',
       'Brent', 'Bromley', 'Camden', 'Croydon', 'Ealing', 'Enfield',
       'Greenwich', 'Hackney', 'Hammersmith & Fulham', 'Haringey',
       'Harrow', 'Havering', 'Hillingdon', 'Hounslow', 'Islington',
       'Kensington & Chelsea', 'Kingston upon Thames', 'Lambeth',
       'Lewisham', 'Merton', 'Newham', 'Redbridge',
       'Richmond upon Thames', 'Southwark', 'Sutton', 'Tower Hamlets',
       'Waltham Forest', 'Wandsworth', 'Westminster', 'Unnamed: 34',
       'Inner London', 'Outer London', 'Unnamed: 37', 'NORTH EAST',
       'NORTH WEST', 'YORKS & THE HUMBER', 'EAST MIDLANDS',
       'WEST MIDLANDS', 'EAST OF ENGLAND', 'LONDON', 'SOUTH EAST',
       'SOUTH WEST', 'Unnamed: 47', 'England'], dtype=object)

In [29]:
#subset rows with NA as ID
clean_properties[clean_properties['ID'].isna()]

London_Borough   ID      Month  Average_price
33       Unnamed: 34  NaN 1995-01-01            NaN
36       Unnamed: 37  NaN 1995-01-01            NaN
46       Unnamed: 47  NaN 1995-01-01            NaN
81       Unnamed: 34  NaN 1995-02-01            NaN
84       Unnamed: 37  NaN 1995-02-01            NaN
...              ...  ...        ...            ...
15348    Unnamed: 37  NaN 2021-08-01            NaN
15358    Unnamed: 47  NaN 2021-08-01            NaN
15393    Unnamed: 34  NaN 2021-09-01            NaN
15396    Unnamed: 37  NaN 2021-09-01            NaN
15406    Unnamed: 47  NaN 2021-09-01            NaN

[963 rows x 4 columns]

In [30]:
#subset for rows without null values
NaNFreeDF1 = clean_properties[clean_properties['Average_price'].notna()]
NaNFreeDF1.head(50)

London_Borough         ID      Month  Average_price
0         City of London  E09000001 1995-01-01    91448.98487
1     Barking & Dagenham  E09000002 1995-01-01    50460.22660
2                 Barnet  E09000003 1995-01-01    93284.51832
3                 Bexley  E09000004 1995-01-01    64958.09036
4                  Brent  E09000005 1995-01-01    71306.56698
5                Bromley  E09000006 1995-01-01    81671.47692
6                 Camden  E09000007 1995-01-01   120932.88810
7                Croydon  E09000008 1995-01-01    69158.16225
8                 Ealing  E09000009 1995-01-01    79885.89069
9                Enfield  E09000010 1995-01-01    72514.69096
10             Greenwich  E09000011 1995-01-01    62300.10169
11               Hackney  E09000012 1995-01-01    61296.52637
12  Hammersmith & Fulham  E09000013 1995-01-01   124902.86020
13              Haringey  E09000014 1995-01-01    76287.56947
14                Harrow  E09000015 1995-01-01    84769.52599
15              Havering  E09000016 1995-01-01    68000.13774
16            Hillingdon  E09000017 1995-01-01    73834.82964
17              Hounslow  E09000018 1995-01-01    72231.70537
18             Islington  E09000019 1995-01-01    92516.48557
19  Kensington & Chelsea  E09000020 1995-01-01   182694.83260
20  Kingston upon Thames  E09000021 1995-01-01    80875.84843
21               Lambeth  E09000022 1995-01-01    67770.98843
22              Lewisham  E09000023 1995-01-01    60491.26109
23                Merton  E09000024 1995-01-01    82070.61330
24                Newham  E09000025 1995-01-01    53539.31919
25             Redbridge  E09000026 1995-01-01    72189.58437
26  Richmond upon Thames  E09000027 1995-01-01   109326.12450
27             Southwark  E09000028 1995-01-01    67885.20344
28                Sutton  E09000029 1995-01-01    71536.97357
29         Tower Hamlets  E09000030 1995-01-01    59865.18995
30        Waltham Forest  E09000031 1995-01-01    61319.44913
31            Wandsworth  E09000032 1995-01-01    88559.04381
32           Westminster  E09000033 1995-01-01   133025.27720
34          Inner London  E13000001 1995-01-01    78251.97650
35          Outer London  E13000002 1995-01-01    72958.79836
37            NORTH EAST  E12000001 1995-01-01    42076.35411
38            NORTH WEST  E12000002 1995-01-01    43958.48001
39    YORKS & THE HUMBER  E12000003 1995-01-01    44803.42878
40         EAST MIDLANDS  E12000004 1995-01-01    45544.52227
41         WEST MIDLANDS  E12000005 1995-01-01    48527.52339
42       EAST OF ENGLAND  E12000006 1995-01-01    56701.59610
43                LONDON  E12000007 1995-01-01    74435.76052
44            SOUTH EAST  E12000008 1995-01-01    64018.87894
45            SOUTH WEST  E12000009 1995-01-01    54705.15790
47               England  E92000001 1995-01-01    53202.77128
48        City of London  E09000001 1995-02-01    82202.77314
49    Barking & Dagenham  E09000002 1995-02-01    51085.77983
50                Barnet  E09000003 1995-02-01    93190.16963
51                Bexley  E09000004 1995-02-01    64787.92069
52                 Brent  E09000005 1995-02-01    72022.26197

In [31]:
# check for missing values with new subset
NaNFreeDF1.count()

London_Borough    14442
ID                14442
Month             14442
Average_price     14442
dtype: int64

In [32]:
# filter the data with NaN values
NaNFreeDF2 = clean_properties.dropna()
NaNFreeDF2.head(50)

London_Borough         ID      Month  Average_price
0         City of London  E09000001 1995-01-01    91448.98487
1     Barking & Dagenham  E09000002 1995-01-01    50460.22660
2                 Barnet  E09000003 1995-01-01    93284.51832
3                 Bexley  E09000004 1995-01-01    64958.09036
4                  Brent  E09000005 1995-01-01    71306.56698
5                Bromley  E09000006 1995-01-01    81671.47692
6                 Camden  E09000007 1995-01-01   120932.88810
7                Croydon  E09000008 1995-01-01    69158.16225
8                 Ealing  E09000009 1995-01-01    79885.89069
9                Enfield  E09000010 1995-01-01    72514.69096
10             Greenwich  E09000011 1995-01-01    62300.10169
11               Hackney  E09000012 1995-01-01    61296.52637
12  Hammersmith & Fulham  E09000013 1995-01-01   124902.86020
13              Haringey  E09000014 1995-01-01    76287.56947
14                Harrow  E09000015 1995-01-01    84769.52599
15              Havering  E09000016 1995-01-01    68000.13774
16            Hillingdon  E09000017 1995-01-01    73834.82964
17              Hounslow  E09000018 1995-01-01    72231.70537
18             Islington  E09000019 1995-01-01    92516.48557
19  Kensington & Chelsea  E09000020 1995-01-01   182694.83260
20  Kingston upon Thames  E09000021 1995-01-01    80875.84843
21               Lambeth  E09000022 1995-01-01    67770.98843
22              Lewisham  E09000023 1995-01-01    60491.26109
23                Merton  E09000024 1995-01-01    82070.61330
24                Newham  E09000025 1995-01-01    53539.31919
25             Redbridge  E09000026 1995-01-01    72189.58437
26  Richmond upon Thames  E09000027 1995-01-01   109326.12450
27             Southwark  E09000028 1995-01-01    67885.20344
28                Sutton  E09000029 1995-01-01    71536.97357
29         Tower Hamlets  E09000030 1995-01-01    59865.18995
30        Waltham Forest  E09000031 1995-01-01    61319.44913
31            Wandsworth  E09000032 1995-01-01    88559.04381
32           Westminster  E09000033 1995-01-01   133025.27720
34          Inner London  E13000001 1995-01-01    78251.97650
35          Outer London  E13000002 1995-01-01    72958.79836
37            NORTH EAST  E12000001 1995-01-01    42076.35411
38            NORTH WEST  E12000002 1995-01-01    43958.48001
39    YORKS & THE HUMBER  E12000003 1995-01-01    44803.42878
40         EAST MIDLANDS  E12000004 1995-01-01    45544.52227
41         WEST MIDLANDS  E12000005 1995-01-01    48527.52339
42       EAST OF ENGLAND  E12000006 1995-01-01    56701.59610
43                LONDON  E12000007 1995-01-01    74435.76052
44            SOUTH EAST  E12000008 1995-01-01    64018.87894
45            SOUTH WEST  E12000009 1995-01-01    54705.15790
47               England  E92000001 1995-01-01    53202.77128
48        City of London  E09000001 1995-02-01    82202.77314
49    Barking & Dagenham  E09000002 1995-02-01    51085.77983
50                Barnet  E09000003 1995-02-01    93190.16963
51                Bexley  E09000004 1995-02-01    64787.92069
52                 Brent  E09000005 1995-02-01    72022.26197

In [33]:
# check for missing values 
NaNFreeDF2.count()

London_Borough    14442
ID                14442
Month             14442
Average_price     14442
dtype: int64

In [34]:
#check that London Borough columns are correct 
NaNFreeDF2['London_Borough'].unique()

array(['City of London', 'Barking & Dagenham', 'Barnet', 'Bexley',
       'Brent', 'Bromley', 'Camden', 'Croydon', 'Ealing', 'Enfield',
       'Greenwich', 'Hackney', 'Hammersmith & Fulham', 'Haringey',
       'Harrow', 'Havering', 'Hillingdon', 'Hounslow', 'Islington',
       'Kensington & Chelsea', 'Kingston upon Thames', 'Lambeth',
       'Lewisham', 'Merton', 'Newham', 'Redbridge',
       'Richmond upon Thames', 'Southwark', 'Sutton', 'Tower Hamlets',
       'Waltham Forest', 'Wandsworth', 'Westminster', 'Inner London',
       'Outer London', 'NORTH EAST', 'NORTH WEST', 'YORKS & THE HUMBER',
       'EAST MIDLANDS', 'WEST MIDLANDS', 'EAST OF ENGLAND', 'LONDON',
       'SOUTH EAST', 'SOUTH WEST', 'England'], dtype=object)

In [35]:
#compare subsets to select desired subset

print(clean_properties.shape)
print(NaNFreeDF1.shape)
print(NaNFreeDF2.shape)

(15408, 4)
(14442, 4)
(14442, 4)


In [36]:
# create list of non-boroughs to drop from selected subset
nonBoroughs = ['Inner London', 'Outer London', 
               'NORTH EAST', 'NORTH WEST', 'YORKS & THE HUMBER', 
               'EAST MIDLANDS', 'WEST MIDLANDS',
              'EAST OF ENGLAND', 'LONDON', 'SOUTH EAST', 
              'SOUTH WEST', 'England']

In [37]:
#filter for row values of London_Borough that are in the nonBoroughs list
NaNFreeDF2[NaNFreeDF2.London_Borough.isin(nonBoroughs)]

London_Borough         ID      Month  Average_price
34           Inner London  E13000001 1995-01-01    78251.97650
35           Outer London  E13000002 1995-01-01    72958.79836
37             NORTH EAST  E12000001 1995-01-01    42076.35411
38             NORTH WEST  E12000002 1995-01-01    43958.48001
39     YORKS & THE HUMBER  E12000003 1995-01-01    44803.42878
...                   ...        ...        ...            ...
15402     EAST OF ENGLAND  E12000006 2021-09-01   327981.68440
15403              LONDON  E12000007 2021-09-01   507253.03520
15404          SOUTH EAST  E12000008 2021-09-01   370886.27240
15405          SOUTH WEST  E12000009 2021-09-01   301327.02220
15407             England  E92000001 2021-09-01   287895.40030

[3852 rows x 4 columns]

In [38]:
# get row values for London_Borough that are not in the nonBoroughs list
NaNFreeDF2[~NaNFreeDF2.London_Borough.isin(nonBoroughs)]

London_Borough         ID      Month  Average_price
0          City of London  E09000001 1995-01-01    91448.98487
1      Barking & Dagenham  E09000002 1995-01-01    50460.22660
2                  Barnet  E09000003 1995-01-01    93284.51832
3                  Bexley  E09000004 1995-01-01    64958.09036
4                   Brent  E09000005 1995-01-01    71306.56698
...                   ...        ...        ...            ...
15388              Sutton  E09000029 2021-09-01   415227.66500
15389       Tower Hamlets  E09000030 2021-09-01   440846.73610
15390      Waltham Forest  E09000031 2021-09-01   505389.18630
15391          Wandsworth  E09000032 2021-09-01   599124.41270
15392         Westminster  E09000033 2021-09-01   965766.35540

[10590 rows x 4 columns]

In [39]:
#reassign rows that are not in nonBoroughs list to new dataframe 
NaNFreeDF2 = NaNFreeDF2[~NaNFreeDF2.London_Borough.isin(nonBoroughs)]

In [40]:
#check out new dataframe of only accurate Boroughs
NaNFreeDF2.head()

London_Borough         ID      Month  Average_price
0      City of London  E09000001 1995-01-01    91448.98487
1  Barking & Dagenham  E09000002 1995-01-01    50460.22660
2              Barnet  E09000003 1995-01-01    93284.51832
3              Bexley  E09000004 1995-01-01    64958.09036
4               Brent  E09000005 1995-01-01    71306.56698

In [41]:
#reassign to final analysis df
df = NaNFreeDF2

**2.6. Visualizing the data**

To visualize the data, why not subset on a particular London Borough? Maybe do a line plot of Month against Average Price?

To limit the number of data points you have, you might want to extract the year from every month value your *Month* column. 

To this end, you *could* apply a ***lambda function***. Your logic could work as follows:
1. look through the `Month` column
2. extract the year from each individual value in that column 
3. store that corresponding year as separate column. 

Whether you go ahead with this is up to you. Just so long as you answer our initial brief: which boroughs of London have seen the greatest house price increase, on average, over the past two decades? 

In [42]:
#one way to subset year from month column
#df['year'] = pd.DatetimeIndex(df['Month']).year
#df.head()

In [43]:
#subset a London Borough
camden_prices = df[df['London_Borough'] == 'Camden']

In [44]:
# or subset years from the dates in the Month column
df['Year'] = df['Month'].apply(lambda t: t.year)

df.tail()

London_Borough         ID      Month  Average_price  Year
15388          Sutton  E09000029 2021-09-01    415227.6650  2021
15389   Tower Hamlets  E09000030 2021-09-01    440846.7361  2021
15390  Waltham Forest  E09000031 2021-09-01    505389.1863  2021
15391      Wandsworth  E09000032 2021-09-01    599124.4127  2021
15392     Westminster  E09000033 2021-09-01    965766.3554  2021

In [45]:
#!! can't get my plot cell to work !!
#create variable for plot method line plot showing Camden's Month and Average Price
#ax = camden_prices.plot(kind ="line", x = "Year", y="Average_price")
#ax.set_ylabel("Price")
#ax.plot("Year", "Average_price")
#plot.show()

In [46]:
#calculate the mean house price for each year by grouping the dataframe by Borough and Year
dfg = df.groupby(by=['London_Borough', 'Year']).mean()
dfg.sample(10)

Average_price
London_Borough       Year               
Southwark            2012  331124.248858
Sutton               2015  322932.138442
Hammersmith & Fulham 2002  302459.055983
Harrow               2002  202680.884917
Islington            2021  680285.846411
Camden               2010  513221.129450
Haringey             2006  263920.937508
Hillingdon           2003  202707.807350
Tower Hamlets        2020  464413.502717
Havering             2011  219558.764817

In [47]:
#reset the index
dfg = dfg.reset_index()
dfg.head()

London_Borough  Year  Average_price
0  Barking & Dagenham  1995   51817.969390
1  Barking & Dagenham  1996   51718.192690
2  Barking & Dagenham  1997   55974.262309
3  Barking & Dagenham  1998   60285.821083
4  Barking & Dagenham  1999   65320.934441

**3. Modeling**

Consider creating a function that will calculate a ratio of house prices, comparing the price of a house in 2018 to the price in 1998.

Consider calling this function create_price_ratio.

You'd want this function to:
1. Take a filter of dfg, specifically where this filter constrains the London_Borough, as an argument. For example, one admissible argument should be: dfg[dfg['London_Borough']=='Camden'].
2. Get the Average Price for that Borough, for the years 1998 and 2018.
4. Calculate the ratio of the Average Price for 1998 divided by the Average Price for 2018.
5. Return that ratio.

Once you've written this function, you ultimately want to use it to iterate through all the unique London_Boroughs and work out the ratio capturing the difference of house prices between 1998 and 2018.

Bear in mind: you don't have to write a function like this if you don't want to. If you can solve the brief otherwise, then great! 

***Hint***: This section should test the skills you acquired in:
- Python Data Science Toolbox - Part One, all modules

In [48]:
#create function to return ratio of house prices in 2018 to 1998
def create_price_ratio(d):
    y1998 = float(d['Average_price'][d['Year']==1998])
    y2018 = float(d['Average_price'][d['Year']==2018])
    ratio = [y2018/y1998]
    return ratio

In [49]:
#call function on one Borough to test
create_price_ratio(dfg[dfg['London_Borough']=='Barking & Dagenham'])

[4.89661861291754]

In [50]:
#create dictionary to store all London Borough ratios
final = {}

In [51]:
# loop through each 'London_Borough' column of our DataFrame dfg.
 
for b in dfg['London_Borough'].unique():
    # subset dfg on 'London_Borough' == b. Each iteration 
    borough = dfg[dfg['London_Borough'] == b]
    # Make a entry in the final dictionary for each result of calling create_price_ratio function with the arrgument borough
    final[b] = create_price_ratio(borough)
# We use the function and incorporate that into a new key of the dictionary 
print(final) 

{'Barking & Dagenham': [4.89661861291754], 'Barnet': [4.358195917538044], 'Bexley': [4.248977046127877], 'Brent': [4.8945544971392865], 'Bromley': [4.094784685333876], 'Camden': [4.935353408884261], 'City of London': [5.30162037758761], 'Croydon': [4.201100280024766], 'Ealing': [4.311450902121834], 'Enfield': [4.263471583495811], 'Greenwich': [4.7630363473291935], 'Hackney': [6.198285561008663], 'Hammersmith & Fulham': [4.13779810193623], 'Haringey': [5.134624964136042], 'Harrow': [4.0591964329643195], 'Havering': [4.325230371335307], 'Hillingdon': [4.2002730803844575], 'Hounslow': [3.976409106143329], 'Islington': [4.844048012802297], 'Kensington & Chelsea': [5.082465066092464], 'Kingston upon Thames': [4.270549521484271], 'Lambeth': [4.957751163514062], 'Lewisham': [5.449221041059686], 'Merton': [4.741273313294603], 'Newham': [5.305390437201879], 'Redbridge': [4.471182006097364], 'Richmond upon Thames': [4.005161895721457], 'Southwark': [5.516485302379378], 'Sutton': [4.1185226085731

In [52]:
#assign the results of calling the DataFrame method on the dictionary final. 
df_ratios = pd.DataFrame(final)

In [53]:
# Call the head() method on this variable to check it out. 
df_ratios.head()

Barking & Dagenham    Barnet    Bexley     Brent   Bromley    Camden  \
0            4.896619  4.358196  4.248977  4.894554  4.094785  4.935353   

   City of London  Croydon    Ealing   Enfield  ...    Merton   Newham  \
0         5.30162   4.2011  4.311451  4.263472  ...  4.741273  5.30539   

   Redbridge  Richmond upon Thames  Southwark    Sutton  Tower Hamlets  \
0   4.471182              4.005162   5.516485  4.118523       4.626701   

   Waltham Forest  Wandsworth  Westminster  
0        5.834756    4.757709     5.353565  

[1 rows x 33 columns]

In [54]:
#transpose and reset the index for better viewing
df_ratios_T = df_ratios.T
df_ratios = df_ratios_T.reset_index()
df_ratios.head()

index         0
0  Barking & Dagenham  4.896619
1              Barnet  4.358196
2              Bexley  4.248977
3               Brent  4.894554
4             Bromley  4.094785

In [55]:
#rename the 'index' column as 'London_Borough', and the '0' column to '2018'.
df_ratios.rename(columns={'index':'Borough', 0:'2018'}, inplace=True)
df_ratios.head()

Borough      2018
0  Barking & Dagenham  4.896619
1              Barnet  4.358196
2              Bexley  4.248977
3               Brent  4.894554
4             Bromley  4.094785

In [56]:
# sort in descending order for top 15 boroughs.
# Make a variable called top15, and assign it the result of calling sort_values() on df_ratios. 
top15 = df_ratios.sort_values(by='2018',ascending=False).head(15)
print(top15)

                 Borough      2018
11               Hackney  6.198286
30        Waltham Forest  5.834756
27             Southwark  5.516485
22              Lewisham  5.449221
32           Westminster  5.353565
24                Newham  5.305390
6         City of London  5.301620
13              Haringey  5.134625
19  Kensington & Chelsea  5.082465
21               Lambeth  4.957751
5                 Camden  4.935353
0     Barking & Dagenham  4.896619
3                  Brent  4.894554
18             Islington  4.844048
10             Greenwich  4.763036


# 4. Conclusion


Given the calculated ratios for each London Bourough price increases from 1998 to 2018, the top 15 listed are the Boroughs with the highest increase over the last two decades.

 Borough                 Increase Ratio from 1998 to 2018
 Hackney                 6.198286
 Waltham Forest          5.834756
 Southwark               5.516485
 Lewisham                5.449221
 Westminster             5.353565
 Newham                  5.305390
 City of London          5.301620
 Haringey                5.134625
 Kensington & Chelsea    5.082465
 Lambeth                 4.957751
 Camden                  4.935353
 Barking & Dagenham      4.896619
 Brent                   4.894554
 Islington               4.844048
 Greenwich               4.763036